### ISSUE TITLE
order manager cannot process multiple frames with orders at once

### ISSUE THREAD

**Issue by anvilvapre** [OpenRA/OpenRA#20941]
@teinarss fyi, not sure if you ever plan to continue work on this.

related to networking.

not a real issue in the current release or bleed. but worth mentioning in case when work continues on support for dynamic order latency.

the client side order manager currently cannot process multiple frames at once as part of the same `ProcessOrders` loop. it can only process a single frame each loop. however  it will always try to process all frames that were received - as part of the same loop.

in release and bleed in practice it turns out that in most cases only ever a single frame is queued. i assume this due to frames being sent in individual tcp packets. in theory perhaps could happen during network congestion.

i was trying to optimize the sending and receiving of orders. when the server receives an order by the client and replies directly with an ack frame - it may reply "too fast" - which can cause a timing issue client side.

i.e. the client crashes at line 238 because it earlier assumed that any received order by the client is still for the current frame index. where actually it may already an be and ack frame that will have the future frame index: current + 1. The ack was being processed before the `NetFrameNumber` could be incremented at the end of `OrderManager.ProcessOrders`.

```Exception of type `System.IO.InvalidDataException`: Attempted to process orders from client 0 for frame 118 on frame 117```

https://github.com/OpenRA/OpenRA/blob/56b5ace109c8fd8221684a24e897156bbcefa4f0/OpenRA.Game/Network/OrderManager.cs#L232-L238

server replying to a client order by sending an ack frame. In the ack frame the frame id is incremented with the order latency which is currently always 1. the client may potentially receive this as part of the same `ProcessOrders` loop.

https://github.com/OpenRA/OpenRA/blob/56b5ace109c8fd8221684a24e897156bbcefa4f0/OpenRA.Game/Server/Server.cs#L907-L918

in future the order manager needs to consider the frame index of queued pending orders and insert incoming orders by frame index. or maintain a map of frame by frame index and use that to determine if its ready for the next frame.

this will likely become also an issue when dynamic latency needs to be supported.

in summary, `IsReadyForNextFrame` may return true incorrectly, due to a future order was already received.

**Comment by teinarss** [comment]
I had a smiliar idea, but with classes that would handle the logic for normal frame orders, decrese and increase order latency orders.
